refactor: add utility for converting Configuration to Bigtable Settings

rahulKQL opened the pull request:
In progress

rahulKQL committed:
refactor: add utility for converting Configuration to Bigtable Settings

## Background
As of right now, We have only one way of user provided configuration porting i.e. `BigtableOptionsFactory#fromConfiguration`. This utlity converts Hbase `Configuration` to `BigtableOptions`, which is present in `bigtable-core-client`.

## What this PR contains
This change introduces two new utility in the same class to convert `Configuration` to veneer's `BigtableDataSettings` & `BigtableTableAdminSettings`.

**Note:** _We need to copy over most of the constants from BigtableOptions, CallOptionsConfig, RetryOptions to `bigtable-hbase` module to maintain seamless user experience(considering some of the users are relying on these constants)._

dmitry-fa commented on bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableOptionsFactory.java:
```diff
@@ -86,7 +123,7 @@ public void testConnectionKeysAreUsed() throws IOException {
-    configuration.setBoolean(BigtableOptionsFactory.BIGTABLE_USE_PLAINTEXT_NEGOTIATION, true);
+    configuration.setBoolean(BIGTABLE_USE_PLAINTEXT_NEGOTIATION, true);
     BigtableOptions options = BigtableOptionsFactory.fromConfiguration(configuration);
 
```
would it make sense to do static import for others BigtableOptionsFactory properties ?

dmitry-fa commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java:
```diff
@@ -550,11 +594,452 @@ private static RetryOptions createRetryOptions(Configuration configuration) {
+    LOG.debug("API Data endpoint host %s.", dataHostOverride);
+
+    String endpoint = dataHostOverride + ":" + portNumber;
+
```
Just a question: you output only 'endpoint host', would it make any sense to output the entire endpoint?

rahulKQL replied:
I couldn't get what you meant by "entire endpoint"

Do you mean to get complete `hostname:port` from a single `configuration.get()`? If yes then it is not possible because Hbase doesn't bind host and port in a single config key.

rahulKQL replied:
Aww, just understood what you mean.. you mean to log complete endpoint. Sorry for the trouble

dmitry-fa replied:
Yes, I mean:
LOG.debug("API Data endpoint host %s.", dataHostOverride);
=>
LOG.debug("endpoint host %s.", endpoint );

But it's not necessary

rahulKQL replied:
Yes, that sounds good to me too. I have updated the log.

dmitry-fa commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java:
```diff
@@ -550,11 +594,452 @@ private static RetryOptions createRetryOptions(Configuration configuration) {
+    if (configuration.getBoolean(ALLOW_NO_TIMESTAMP_RETRIES_KEY, false)) {
+      throw new UnsupportedOperationException("Retries without Timestamp does not support yet.");
+    }
+
```
Retries without Timestamp does not support yet.
=>
No Timestamp retries are not supported yet.

Where do you check, that timestamp is not defined?

rahulKQL replied:
> Where do you check, that timestamp is not defined?

The timestamp logic is in the bigtable-core module([here](https://github.com/googleapis/java-bigtable-hbase/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java#L224-L235)). But with veneer client this logic is pushed back to gax, and we do not have any control to enable this check.

dmitry-fa replied:
I'm asking because you throw an exception if ALLOW_NO_TIMESTAMP_RETRIES_KEY == false, but you do not check if TIMESTAMP == null. 
If you think this is ok, I'm ok )

rahulKQL replied:
The timestamp would be picked while a user creates `Mutation` object.

The core-client allows mutations without a timestamp. With veneer client, each row mutation must have a timestamp(either client-side or server-side) attached to it. 

I hope I am clear this time, if not then I can explain more.

dmitry-fa replied:
Oh, yes. I see. `buildIdempotentRetrySettings()` checks not the retry, but the settings. And it will fail if one tries to create settings which allow no timestamp retries. 
Thanks for explanation.

I have no other questions

dmitry-fa commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java:
```diff
@@ -550,11 +594,452 @@ private static RetryOptions createRetryOptions(Configuration configuration) {
+    LOG.debug("API Data endpoint host %s.", dataHostOverride);
+
+    String endpoint = dataHostOverride + ":" + portNumber;
+
```
question: would it make sense to print out the entire endpoint, not only host?

dmitry-fa commented on bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableOptionsFactory.java:
```diff
@@ -176,4 +213,238 @@ public void testLongOperationsTimeouts() throws IOException {
+    configuration.set(BigtableOptionsFactory.CUSTOM_USER_AGENT_KEY, userAgent);
+    configuration.setBoolean(BIGTABLE_USE_PLAINTEXT_NEGOTIATION, true);
+    configuration = BigtableConfiguration.withCredentials(configuration, credentials);
+
```
Here and below: you can add static import of other BigtableOptionsFactory properties

dmitry-fa commented on bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableOptionsFactory.java:
```diff
@@ -86,7 +123,7 @@ public void testConnectionKeysAreUsed() throws IOException {
-    configuration.setBoolean(BigtableOptionsFactory.BIGTABLE_USE_PLAINTEXT_NEGOTIATION, true);
+    configuration.setBoolean(BIGTABLE_USE_PLAINTEXT_NEGOTIATION, true);
     BigtableOptions options = BigtableOptionsFactory.fromConfiguration(configuration);
 
```
static imports for BigtableOptionsFactory properties

rahulKQL replied:
Thanks for the review!!... Statically imported constants.

rahulKQL committed:
static imported constants and updated log messages

dmitry-fa commented:
LGTM

rahulKQL committed:
chore: addressing feedback comment

- hostname are only updated if overridden.
- isChannelPoolCachingEnabled() can confirm cases for channel pooling.
- all configurations are only updated if user has override them else this class defaults to Veneer default values.
